Test plan updater

evmiguel opened the pull request:
This is the PR for the Test Plan Updater Modal. To see it locally,

- Start development server at localhost:3000: `yarn dev`
- Log into the app as an admin
- Navigate to the Test Queue
- Pick a report and press  the `Update Test Plan Version` to see the modal
- Try creating a new test plan report

----
Open questions:

- Which versions should be shown in the dropdown?
- What should happen when the admin does not choose to delete the old report?

cc: @richnoah @s3ththompson

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdater.jsx:
```diff
@@ -0,0 +1,436 @@
+                <title>Test Plan Updater | ARIA-AT</title>
+            </Helmet>
+            <h1>Test Plan Updater</h1>
+            <h2>Explanation</h2>
```
Shall we remove this UI which is no longer used?

alflennik commented on client/routes/index.js:
```diff
@@ -56,6 +57,14 @@ export default [
+                <Route component={TestPlanUpdater} />
+            </ConfirmAuth>
+        )
+    },
```
This might not be needed anymore, right?

evmiguel replied:
Right. I've removed it.

evmiguel commented on client/index.js:
```diff
@@ -8,7 +8,37 @@ import App from './components/App';
+                    }
+                }
+            }
+        }
```
Works!

alflennik replied:
This is minor, but do you want to put the `{ merge: false }` and `{merge: true }` parts on one line? That'll shorten this code by 8 lines 😈

richnoah commented:
@s3ththompson @evmiguel I don't see the need to downgrade as such a high priority.  As explained there are bigger implications of reverting a test plan version.  I would be fine with just the most recent version being shown.  If I have selected the most current version and then go back into the updater is there a check if the selected version is the most current?

@evmiguel I would say it is bad form to have a pop-up above a modal.  Possibly using a toast message that an admin would have to confirm

s3ththompson commented:
> If I have selected the most current version and then go back into the updater is there a check if the selected version is the most current?

If I understand correctly @richnoah, I think this is a great approach. So instead of a dropdown, the test plan updater only shows the most recent test version. If the test plan is already the recent version, there would be no "Update Test Plan" button in the test queue. Would that be feasible @evmiguel?

> I would say it is bad form to have a pop-up above a modal.

@richnoah I don't know if we have that kind of toast component in our component library. Perhaps it's enough to have a warning and to make the button red?

evmiguel commented:
@s3ththompson @richnoah @isaacdurazo:

I've removed the select dropdown to make sure that we can only update on the most recent version. Reports that are already on the recent version do not have the update option.

I added alerts to the modal when the old report is being deleted. It feels a bit awkward to me, but I'd like to get y'alls opinions.

alflennik commented on client/components/TestQueueRow/index.jsx:
```diff
@@ -120,6 +127,24 @@ const TestQueueRow = ({
+                    variant="outline-secondary"
+                >
+                    {' '}
+                    Update Test Plan Version
```
Since it's the "Test Plan Updater" now I feel like this button can be called "Update Test Plan" which would also make it more compact. I like the new styling. Good job picking its location, so far everyone's been happy with it.

Also could you remove the `{' '}`?

alflennik commented on client/components/TestQueueRow/TestQueueRow.css:
```diff
@@ -103,6 +103,10 @@ button.more-actions:active {
 
+.updater-button {
+    max-width: 30%;
+}
```
At a larger screen size this is solving the problem of making the button a reasonable size instead of stretching across the entire cell - but on smaller screens the button shrinks too fast!

```
.btn.updater-button {
    width: initial;
}
```

This will accomplish the same task and work on smaller screens too.

I had to add the `.btn` part to override another style.

alflennik commented on client/components/LoadingSpinner/LoadingSpinner.jsx:
```diff
@@ -0,0 +1,46 @@
+                className
+                    ? `spinner-container ${className}`
+                    : 'spinner-container'
+            }
```
A nice way to get this on one line would be 

```
className={`spinner-container ${className}`}
```

And change the default className to `className = ''`  on line 5.

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.jsx:
```diff
@@ -0,0 +1,541 @@
+                                    deletionNote = (
+                                        <>
+                                            Note that {testsToDelete.length}{' '}
+                                            tests differ between the old and new
```
Could we just make it "test" in the case that there's just one test which can't be copied?

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.jsx:
```diff
@@ -0,0 +1,541 @@
+            <Form.Check.Label className="delete-label">
+                <FontAwesomeIcon icon={faTrashAlt} className="trash-icon" />{' '}
+                Delete old Test Plan
+            </Form.Check.Label>
```
Clicking the text didn't check the checkbox. I guess you might need to either put the checkbox inside the label or add an `id` prop to the checkbox and a `for` prop for the label.

alflennik reviewed (commented):
I'm really proud of what we've accomplished with this feature. At this point we've left our previous sketchy version so far behind, it's amazing. I think what we have here is really user-friendly, and it's really well implemented too. Great job!

howard-e commented on client/components/TestQueueRow/index.jsx:
```diff
@@ -120,6 +127,24 @@ const TestQueueRow = ({
+                    Update Test Plan Version
+                </Button>
+            ) : (
+                ''
```
nitpick and completely optional: I'm not sure if the same happens with hooks as it did with `React.Component` but doing this could create a `React.Text` node I believe.
It definitely has no impact on performance like this of course but the standard for preventing component rendering also seems to be defaulting to `null`.

Could also just avoid the ternary altogether and do something like:
```jsx
const updateTestPlanVersionButton =
            isAdmin && latestTestPlanVersion.length === 0 && (
                <Component ... />
            );
```

howard-e commented on client/components/TestQueueRow/index.jsx:
```diff
@@ -120,6 +127,24 @@ const TestQueueRow = ({
+                    className="updater-button"
+                    onClick={() => setShowTestPlanUpdaterModal(true)}
+                    size="sm"
+                    variant="outline-secondary"
```
My initial feeling about this visually was that I didn't expect it could be interacted with because of the variant. Would be curious to know if anyone else had the same impression. But if I'm alone in my thoughts here, then no problem!

howard-e reviewed (commented):
Impressive stuff @evmiguel and happy to see this getting in there! 🙂

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.css:
```diff
@@ -0,0 +1,80 @@
+    display: grid;
+    grid-template-columns: 1fr 1fr;
+    width: 100%;
+}
```
I noticed the button text was wrapping, and I think it should be possible to keep the button text on one line. I think the css can be simplified slightly! And the HTML can be simplified too - you can remove the divs with the classes which are no longer needed.

```css
.side-button-container {
    display: grid;
    grid-template-columns: auto auto;
    grid-gap: 0.5em;
}
/* Can be removed */
/*
.side-button-container.no-backup {
    grid-column: 2/3;
}

.submit-button {
    width: 100%;
    justify-self: end;
}

.submit-buttons-row {
    display: grid;
    grid-template-columns: 1fr 1fr;
    width: 100%;
}
*/
```

evmiguel replied:
<img width="726" alt="Screen Shot 2022-03-22 at 3 10 33 PM" src="https://user-images.githubusercontent.com/2789098/159557584-441d412c-d1d2-48c6-91b5-fe90952e6fe4.png">

The backup checkbox gets rendered conditionally, so I need the styles. to keep the checkbox on the left and the two buttons on the right.

alflennik replied:
Ah, I get it, there's some styles and HTML to keep the checkbox above the buttons. Ok, so the main thing I wanted to address was the fact the button didn't have enough space and was wrapping. I would say, if you could figure out a way to fix that, that's the main thing motivating my suggestion. I'm happy to look at it together if that helps.

evmiguel replied:
@alflennik I've addressed the text wrapping. It's on sandbox. Mind taking a look?

alflennik replied:
Looks great.

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.jsx:
```diff
@@ -0,0 +1,555 @@
+            setCloseButton(true);
+        } else {
+            setAlertCompletion({
+                message: 'Old test plan will be deleted.',
```
Could we do "The old test plan will be deleted."?

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.jsx:
```diff
@@ -0,0 +1,555 @@
+                <Modal.Footer className="test-plan-updater-footer">
+                    {(!closeButton && (
+                        <Button variant="secondary" onClick={deleteOldTestPlan}>
+                            Next
```
Could this be red and say "Delete"? That's assuming this suggestion doesn't conflict with any specific direction you've already received about how to implement this.

alflennik commented on client/components/TestPlanUpdater/TestPlanUpdaterModal.jsx:
```diff
@@ -0,0 +1,555 @@
+                                            }}
+                                            checked={backupChecked}
+                                        />
+                                        Backup Test Plan before Update
```
What would you think about changing this to "Retain old test plan"?

Or if the copy is pretty-much decided already, could you change this to "Back up test plan before update"? I do prefer the word "retain" if that's possible.

evmiguel replied:
I have no preference. @s3ththompson @isaacdurazo what do you think is the preferred wording?

alflennik replied:
I do think the existing test will work, thanks for updating it.

alflennik reviewed (approved):
Looking solid! I couldn't really find any notes for the parts of the code I've already reviewed, I'm really happy with it. I did have a couple suggested changes inside the brand new additions, which hopefully won't be too involved. Almost there!
